Source/WebCore/loader/HistoryController.cpp

```diff
@@ -877,8 +880,8 @@ void HistoryController::updateBackForwardListClippedAtTarget(bool doClip)
     if (frame->loader().protectedDocumentLoader()->urlForHistory().isEmpty())
         return;
 
-    Ref mainFrame = frame->mainFrame();
-    Ref topItem = mainFrame->checkedHistory()->createItemTree(page->historyItemClient(), *frame, doClip);
+    Ref rootFrame = frame->rootFrame();
+    Ref topItem = rootFrame->checkedHistory()->createItemTree(page->historyItemClient(), *frame, doClip);
```

Comment:
I'm not sure that switching from the main frame to the root frame is conceptually right.

Reply:
Same as above - I don’t think there is any reason to construct history items for root frames that are not parents of the navigating frame.